src/segmentmytiff/utils/models.py

```diff
@@ -0,0 +1,75 @@
+import torch
+from torch import nn
+
+
+class DoubleConv(nn.Module):
+    def __init__(self, in_channels, out_channels):
+        super().__init__()
+        self.conv_op = nn.Sequential(
+            nn.Conv2d(in_channels, out_channels, kernel_size=3, padding=1),
+            nn.ReLU(inplace=True),
+            nn.Conv2d(out_channels, out_channels, kernel_size=3, padding=1),
+            nn.ReLU(inplace=True)
+        )
+
+    def forward(self, x):
+        return self.conv_op(x)
+
+
+class DownSample(nn.Module):
+    def __init__(self, in_channels, out_channels):
+        super().__init__()
+        self.conv = DoubleConv(in_channels, out_channels)
+        self.pool = nn.MaxPool2d(kernel_size=2, stride=2)
+
+    def forward(self, x):
+        down = self.conv(x)
+        p = self.pool(down)
+
+        return down, p
+
+
+class UpSample(nn.Module):
+    def __init__(self, in_channels, out_channels):
+        super().__init__()
+        self.up = nn.ConvTranspose2d(in_channels, in_channels // 2, kernel_size=2, stride=2)
+        self.conv = DoubleConv(in_channels, out_channels)
+
+    def forward(self, x1, x2):
+        x1 = self.up(x1)
+        x = torch.cat([x1, x2], 1)
+        return self.conv(x)
+
+
+class UNet(nn.Module):
+    def __init__(self, in_channels, num_classes):
+        super().__init__()
+        self.down_convolution_1 = DownSample(in_channels, 64)
+        self.down_convolution_2 = DownSample(64, 128)
+        self.down_convolution_3 = DownSample(128, 256)
+        self.down_convolution_4 = DownSample(256, 512)
```

Comment:
Is there a reason to have 4 convolution always? I am thinking if it might be worth to generalize this code to get the number of convolutions as a parameter, and generate each of these layers in a loop.

Reply:
No good reason. I just took the architecture from the tutorial on which this approach was based. That's probably the reason why training was successful on the first try already.

Absolutely a good idea to parametrize this part of the architecture if we want to search for a better model later. 👍